### PR TITLE
[Proposal] Add Stages

### PR DESCRIPTION
**Pull request by alanpoulain**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

As I've done with GraphQL (https://github.com/api-platform/core/pull/2959), I would like to do the same for the REST part.

The aim is to create **Stage** services.

They will be responsible for all the stages (or steps) API Platform is doing to resolve a request.

They will be independent of the request and will be used by the event listeners.

A lot of benefits:
- instead of using events and having issues with priorities (https://api-platform.com/docs/core/events/), the services could be decorated instead;
- since they will be independent of events and the request, they could be used in different contexts (I'm thinking of Laravel for instance).

Right now I've only done the read stage. Behat tests are passing.

WDYT @api-platform/core-team?

### CODE REVIEW

File: src/Stage/ReadStage.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        $this->itemDataProvider = $itemDataProvider;
+        $this->subresourceDataProvider = $subresourceDataProvider;
+        $this->identifierConverter = $identifierConverter;
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(array $attributes, array $parameters, ?array $filters, string $method, array $normalizationContext)
```
Comment (dunglas) [inline src/Stage/ReadStage.php]: Instead of this long list of unstructured and untyped parameters, shouldn't we introduce a new `ApiRequest` or `ApiQuery`/`ApiCommand` class(es), that we'll construct from the (Symfony/Laravel/PSR-7) HTTP request, or from the GraphQL query, or from any other source of request regarless of the protocol. It will allow to have a more strict typing, and will improve the evolvability of API Platform (in case we need to add more data over the time).

Basically, the responsibility of the Symfony Event Listener, or of the Laravel middleware, or of the GraphQL resolver would be to create this `ApiRequest` instance.

File: src/Stage/ReadStage.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            $context[IdentifierConverterInterface::HAS_IDENTIFIER_CONVERTER] = true;
+        }
+
+        try {
+            $identifiers = $this->extractIdentifiers($parameters, $attributes);
+
+            if (isset($attributes['item_operation_name'])) {
+                $data = $this->getItemData($identifiers, $attributes, $context);
+            } elseif (isset($attributes['subresource_operation_name'])) {
+                // Legacy
```
Comment (dunglas) [inline src/Stage/ReadStage.php]: Why legacy? It's a valid use case to not inject a `SubResourceDataPprovider` (especially if you use the library directly, without the Symfony bridge).

File: src/Stage/ReadStageInterface.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Stage;
+
+/**
+ * Retrieves data from the applicable data provider and sets it as a request parameter called data.
```
Comment (dunglas) [inline src/Stage/ReadStageInterface.php]: Outdated

File: src/EventListener/ReadListener.php [hunk 4]
```diff
@@ -60,61 +57,33 @@ public function __construct(CollectionDataProviderInterface $collectionDataProvi
...
+        }
+
+        if ($normalizationContext) {
+            // Builtin data providers are able to use the serialization context to automatically add join clauses
+            $request->attributes->set('_api_normalization_context', $normalizationContext);
         }
 
         $request->attributes->set('data', $data);
-        $request->attributes->set('previous_data', \is_object($data) ? clone $data : $data);
+        $request->attributes->set('previous_data', \is_object($data) && (new \ReflectionClass(\get_class($data)))->isCloneable() ? clone $data : $data);
```
Comment (dunglas) [inline src/EventListener/ReadListener.php]: This should be backported on 2.4 as a bugfix.

File: src/EventListener/ReadListener.php [hunk 5]
```diff
@@ -35,21 +34,19 @@
...
-    use OperationDataProviderTrait;
-    use ToggleableOperationAttributeTrait;
-
     public const OPERATION_ATTRIBUTE_KEY = 'read';
 
     private $serializerContextBuilder;
+    private $readStage;
 
-    public function __construct(CollectionDataProviderInterface $collectionDataProvider, ItemDataProviderInterface $itemDataProvider, SubresourceDataProviderInterface $subresourceDataProvider = null, SerializerContextBuilderInterface $serializerContextBuilder = null, IdentifierConverterInterface $identifierConverter = null, ResourceMetadataFactoryInterface $resourceMetadataFactory = null)
+    public function __construct(CollectionDataProviderInterface $collectionDataProvider, ItemDataProviderInterface $itemDataProvider, SubresourceDataProviderInterface $subresourceDataProvider = null, SerializerContextBuilderInterface $serializerContextBuilder = null, IdentifierConverterInterface $identifierConverter = null, ResourceMetadataFactoryInterface $resourceMetadataFactory = null, ReadStageInterface $readStage = null)
```
Comment (dunglas) [inline src/EventListener/ReadListener.php]: Shouldn't the listener depend only of this new service? And this new service be the first parameter, and passing all other services been deprecated?

File: src/Stage/ReadStageInterface.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+interface ReadStageInterface
+{
+    public const OPERATION_ATTRIBUTE_KEY = 'read';
+
+    /**
+     * @return object|iterable|null
+     */
+    public function apply(array $attributes, array $parameters, ?array $filters, string $method, array $normalizationContext);
```
Comment (dunglas) [inline src/Stage/ReadStageInterface.php]: `apply` isn't very descriptive (I'm not a fond of "stage" too btw :D). Can't we more explicit names?

File: src/Stage/ReadStage.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+use ApiPlatform\Core\Identifier\IdentifierConverterInterface;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ToggleableOperationAttributeTrait;
+
+/**
+ * Retrieves data from the applicable data provider and sets it as a request parameter called data.
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+final class ReadStage implements ReadStageInterface
```
Comment (dunglas) [inline src/Stage/ReadStage.php]: Why isn't this class also used by the GraphQL resolver?

File: src/Stage/ReadStageInterface.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Stage;
+
+/**
+ * Retrieves data from the applicable data provider and sets it as a request parameter called data.
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+interface ReadStageInterface
```
Comment (dunglas) [inline src/Stage/ReadStageInterface.php]: I wonder if this new concept is really necessary. Does this logic belong to the DataProvider? Or should we create a `ApiRequestFactory` class instead (see my previous comment?). I'm not sure yet, I need to think about it, but I've the feeling that we're introducing unnecessary indirection layers.

File: src/Stage/ReadStage.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+use ApiPlatform\Core\DataProvider\SubresourceDataProviderInterface;
+use ApiPlatform\Core\Exception\InvalidIdentifierException;
+use ApiPlatform\Core\Exception\NotFoundException;
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\Identifier\IdentifierConverterInterface;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ToggleableOperationAttributeTrait;
+
+/**
+ * Retrieves data from the applicable data provider and sets it as a request parameter called data.
```
[the remark this answers is not in the document]
Comment (alanpoulain) [reply src/Stage/ReadStage.php]: You think? It's basically what this stage does (excepted the set part which I can remove).

File: src/Stage/ReadStage.php (hunk 1 above)
Comment (alanpoulain) [reply src/Stage/ReadStage.php]: I've always wondered if it would be interesting to introduce this kind of class instead of having a lot of parameters and a `context` array with more things inside :sweat_smile:
I'm :+1: for it.

File: src/Stage/ReadStage.php (hunk 2 above)
Comment (alanpoulain) [reply src/Stage/ReadStage.php]: Not my code here but I can remove the comment. The `SubresourceData` is mandatory in the GraphQL read stage.

File: src/Stage/ReadStageInterface.php (hunk 3 above)
Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: See above.

File: src/EventListener/ReadListener.php (hunk 4 above)
Comment (alanpoulain) [reply src/EventListener/ReadListener.php]: This is not really a bugfix. It's because now this method is *always* called whereas before it was not called if the operation was a collection.

File: src/EventListener/ReadListener.php (hunk 5 above)
Comment (alanpoulain) [reply src/EventListener/ReadListener.php]: You're right.

File: src/Stage/ReadStageInterface.php (hunk 6 above)
Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: Which one do you have in mind? Something like `read`?
I know you would say that :smile: Naming is really hard and subjective. "Stage" was interesting because it was not used and its meaning was alright: it's a "step in a process".

File: src/Stage/ReadStage.php (hunk 7 above)
Comment (alanpoulain) [reply src/Stage/ReadStage.php]: I would like to merge both too but sadly they are very different. I don't think it will be doable easily.

File: src/Stage/ReadStageInterface.php (hunk 8 above)
Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: I think their responsibility is clear:
- allow or not to apply the stage
- extract, collect and normalize the right parameters (mainly extract the identifiers and build the (de)normalization context)
- call the good "solver"
We *could* add the second point inside an `ApiRequestFactory`. But IMO it will not necessarily be a good thing. It will have a lot of conditional clauses and could quickly become a monster.
And I don't know where we could put the other parts. Maybe having a `ChainDataProvider` and modify the `ChainDataPersister` to choose the right method (`persist` or `remove` to call based on the `ApiRequest`)?

Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: Also IMHO, it's easier for the user to understand and decorate this kind of service than to use the data provider / persister and the serializer.

File: src/EventListener/ReadListener.php (hunk 4 above)
Comment (dunglas) [reply src/EventListener/ReadListener.php]: An issue has been reported for non-cloneable objects IIRC

File: src/Stage/ReadStageInterface.php (hunk 8 above)
Comment (dunglas) [reply src/Stage/ReadStageInterface.php]: I’m really not sure about the last point! It’s more concepts to learn with no clear responsibilities.

File: src/Stage/ReadStageInterface.php (hunk 6 above)
Comment (dunglas) [reply src/Stage/ReadStageInterface.php]: Stage sounds like “util” or “service” to me: too generic and not very descriptive

File: src/Stage/ReadStage.php (hunk 7 above)
Comment (dunglas) [reply src/Stage/ReadStage.php]: But then it doesn’t fix the issue of having good extensions points working both for REST and GraphQL. Which is one of the goals.

Comment (alanpoulain) [reply src/Stage/ReadStage.php]: Why not having different ones? The need to use both REST and GraphQL is not common. It would just mean having two different service names to use.

File: src/EventListener/ReadListener.php (hunk 4 above)
Comment (alanpoulain) [reply src/EventListener/ReadListener.php]: In this case, without this change, the tests wouldn't pass.

File: src/Stage/ReadStageInterface.php (hunk 8 above)
Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: I guess it depends of the developer and their way to think.
But a lot of developers are using the events already to add more logic so the need is there IMO.

File: src/Stage/ReadStageInterface.php (hunk 6 above)
Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: But it is something kind of generic. It's a step or an operation. How would you call the `read`, `serialize`, `validate`, etc. things?

Comment (bendavies) [reply src/Stage/ReadStageInterface.php]: Make it a callable? `__invoke`?

Comment (alanpoulain) [reply src/Stage/ReadStageInterface.php]: Good idea! WDYT @dunglas?

**Comment by dunglas** [comment]
I need to find some time to dive deeper into this. I'm still not convinced that we need this new layer of complexity. I also don't like the fact that this new class isn't used by the GraphQL subsystem.

File: src/Stage/ReadStage.php (hunk 7 above)
Comment (Goutte) [reply src/Stage/ReadStage.php]: To me it's very common, implementing both.  That's actually what drove me to API Platform in the first place, I thought it would provide a(nother) way to DRY my Controllers and Resolvers.  If you feel it's too much work for this MR, that's of course your prerogative, but please don't sacrifice forever the goal of unification.  Also, `Stage` feels okay ; it fits, and I can't find anything I deem more suitable.

Comment (alanpoulain) [reply src/Stage/ReadStage.php]: We could introduce stage for both REST and GraphQL, but the service will be an empty shell with practically no code IMO. It will be just there to be decorated.
